# **<u> Chargement des données : </u>**

In [ ]:
import pandas as pd
import plotly.express as px

df = pd.read_csv(
    "SeoulBikeData.csv",
    encoding="cp1252"
)

df.head()
df.info()

# **<u> Nettoyage des données : </u>**

### df.shape
### df.info() - On cherche à comprendre quelle est la structure du Dataframe
### df.head() - On cherche à comprendre à qoui ressemblent les données, avoir une première inspection visuelle
### df.isna().sum() - On cherche à comprendre combien de donées sont manquantes?
### df.nunique() - On cherche à comprendre combien existe-t-il de valeurs différentes dans chaque colonne
### df.describe() - On cherche à comprendre comment se comportent les variable numériques
### df.duplicated().sum() - Vérifier combine on a des doublons

In [2]:
df.shape

(8760, 14)

In [3]:
df.isna().sum()

Date                         0
Rented Bike Count            0
Hour                         0
Temperature(°C)              0
Humidity(%)                  0
Wind speed (m/s)             0
Visibility (10m)             0
Dew point temperature(°C)    0
Solar Radiation (MJ/m2)      0
Rainfall(mm)                 0
Snowfall (cm)                0
Seasons                      0
Holiday                      0
Functioning Day              0
dtype: int64

In [4]:
df.nunique()

Date                          365
Rented Bike Count            2166
Hour                           24
Temperature(°C)               546
Humidity(%)                    90
Wind speed (m/s)               65
Visibility (10m)             1789
Dew point temperature(°C)     556
Solar Radiation (MJ/m2)       345
Rainfall(mm)                   61
Snowfall (cm)                  51
Seasons                         4
Holiday                         2
Functioning Day                 2
dtype: int64

In [5]:
df.describe()

,Rented Bike Count,Hour,Temperature(°C),Humidity(%),Wind speed (m/s),Visibility (10m),Dew point temperature(°C),Solar Radiation (MJ/m2),Rainfall(mm),Snowfall (cm)
count,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000
mean,704.602055,11.500000,12.882922,58.226256,1.724909,1436.825799,4.073813,0.569111,0.148687,0.075068
std,644.997468,6.922582,11.944825,20.362413,1.036300,608.298712,13.060369,0.868746,1.128193,0.436746
min,0.000000,0.000000,-17.800000,0.000000,0.000000,27.000000,-30.600000,0.000000,0.000000,0.000000
25%,191.000000,5.750000,3.500000,42.000000,0.900000,940.000000,-4.700000,0.000000,0.000000,0.000000
50%,504.500000,11.500000,13.700000,57.000000,1.500000,1698.000000,5.100000,0.010000,0.000000,0.000000
75%,1065.250000,17.250000,22.500000,74.000000,2.300000,2000.000000,14.800000,0.930000,0.000000,0.000000
max,3556.000000,23.000000,39.400000,98.000000,7.400000,2000.000000,27.200000,3.520000,35.000000,8.800000


In [6]:
df.duplicated().sum()

np.int64(0)

In [7]:
# Convertir la date (format Seoul = jour/mois/année)
df["Date"] = pd.to_datetime(df["Date"], format="%d/%m/%Y")
df["Day_of_week"] = df["Date"].dt.day_name()
df["Month"] = df["Date"].dt.month
df["Is_weekend"] = df["Date"].dt.dayofweek >= 5

In [8]:
df.describe()

,Date,Rented Bike Count,Hour,Temperature(°C),Humidity(%),Wind speed (m/s),Visibility (10m),Dew point temperature(°C),Solar Radiation (MJ/m2),Rainfall(mm),Snowfall (cm),Month
count,8760,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000
mean,2018-05-31 23:59:59.999999744,704.602055,11.500000,12.882922,58.226256,1.724909,1436.825799,4.073813,0.569111,0.148687,0.075068,6.526027
min,2017-12-01 00:00:00,0.000000,0.000000,-17.800000,0.000000,0.000000,27.000000,-30.600000,0.000000,0.000000,0.000000,1.000000
25%,2018-03-02 00:00:00,191.000000,5.750000,3.500000,42.000000,0.900000,940.000000,-4.700000,0.000000,0.000000,0.000000,4.000000
50%,2018-06-01 00:00:00,504.500000,11.500000,13.700000,57.000000,1.500000,1698.000000,5.100000,0.010000,0.000000,0.000000,7.000000
75%,2018-08-31 00:00:00,1065.250000,17.250000,22.500000,74.000000,2.300000,2000.000000,14.800000,0.930000,0.000000,0.000000,10.000000
max,2018-11-30 00:00:00,3556.000000,23.000000,39.400000,98.000000,7.400000,2000.000000,27.200000,3.520000,35.000000,8.800000,12.000000
std,NaN,644.997468,6.922582,11.944825,20.362413,1.036300,608.298712,13.060369,0.868746,1.128193,0.436746,3.448048


In [9]:
df.groupby("Functioning Day")["Rented Bike Count"].describe()

,count,mean,std,min,25%,50%,75%,max
Functioning Day,,,,,,,,
No,295.0,0.000000,0.000000,0.0,0.0,0.0,0.0,0.0
Yes,8465.0,729.156999,642.351166,2.0,214.0,542.0,1084.0,3556.0


In [10]:
# Isoler les jours de fermeture jour férié  
df_active = df[df["Functioning Day"] == "Yes"].copy()
df.describe()

,Date,Rented Bike Count,Hour,Temperature(°C),Humidity(%),Wind speed (m/s),Visibility (10m),Dew point temperature(°C),Solar Radiation (MJ/m2),Rainfall(mm),Snowfall (cm),Month
count,8760,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000,8760.000000
mean,2018-05-31 23:59:59.999999744,704.602055,11.500000,12.882922,58.226256,1.724909,1436.825799,4.073813,0.569111,0.148687,0.075068,6.526027
min,2017-12-01 00:00:00,0.000000,0.000000,-17.800000,0.000000,0.000000,27.000000,-30.600000,0.000000,0.000000,0.000000,1.000000
25%,2018-03-02 00:00:00,191.000000,5.750000,3.500000,42.000000,0.900000,940.000000,-4.700000,0.000000,0.000000,0.000000,4.000000
50%,2018-06-01 00:00:00,504.500000,11.500000,13.700000,57.000000,1.500000,1698.000000,5.100000,0.010000,0.000000,0.000000,7.000000
75%,2018-08-31 00:00:00,1065.250000,17.250000,22.500000,74.000000,2.300000,2000.000000,14.800000,0.930000,0.000000,0.000000,10.000000
max,2018-11-30 00:00:00,3556.000000,23.000000,39.400000,98.000000,7.400000,2000.000000,27.200000,3.520000,35.000000,8.800000,12.000000
std,NaN,644.997468,6.922582,11.944825,20.362413,1.036300,608.298712,13.060369,0.868746,1.128193,0.436746,3.448048


In [11]:
df['Date']

0      2017-12-01
1      2017-12-01
2      2017-12-01
3      2017-12-01
4      2017-12-01
          ...    
8755   2018-11-30
8756   2018-11-30
8757   2018-11-30
8758   2018-11-30
8759   2018-11-30
Name: Date, Length: 8760, dtype: datetime64[ns]

In [12]:
df_active.shape

(8465, 17)

In [13]:
df_active.describe()

,Date,Rented Bike Count,Hour,Temperature(°C),Humidity(%),Wind speed (m/s),Visibility (10m),Dew point temperature(°C),Solar Radiation (MJ/m2),Rainfall(mm),Snowfall (cm),Month
count,8465,8465.000000,8465.000000,8465.000000,8465.000000,8465.000000,8465.000000,8465.000000,8465.000000,8465.00000,8465.000000,8465.000000
mean,2018-05-28 09:14:23.744831744,729.156999,11.507029,12.771057,58.147194,1.725883,1433.873479,3.944997,0.567868,0.14912,0.077685,6.438984
min,2017-12-01 00:00:00,2.000000,0.000000,-17.800000,0.000000,0.000000,27.000000,-30.600000,0.000000,0.00000,0.000000,1.000000
25%,2018-02-27 00:00:00,214.000000,6.000000,3.000000,42.000000,0.900000,935.000000,-5.100000,0.000000,0.00000,0.000000,3.000000
50%,2018-05-28 00:00:00,542.000000,12.000000,13.500000,57.000000,1.500000,1690.000000,4.700000,0.010000,0.00000,0.000000,6.000000
75%,2018-08-24 00:00:00,1084.000000,18.000000,22.700000,74.000000,2.300000,2000.000000,15.200000,0.930000,0.00000,0.000000,9.000000
max,2018-11-30 00:00:00,3556.000000,23.000000,39.400000,98.000000,7.400000,2000.000000,27.200000,3.520000,35.00000,8.800000,12.000000
std,NaN,642.351166,6.920899,12.104375,20.484839,1.034281,609.051229,13.242399,0.868245,1.12554,0.444063,3.452362


In [14]:
### Renomer les colonnes :
df_active.columns = [
    "date", "rented_bike_count", "hour", "temperature",
    "humidity", "wind_speed", "visibility", "dew_point",
    "solar_radiation", "rainfall", "snowfall",
    "season", "holiday", "functioning_day",
    "day_of_week", "month", "is_weekend"
]

In [15]:
df_active.describe()

,date,rented_bike_count,hour,temperature,humidity,wind_speed,visibility,dew_point,solar_radiation,rainfall,snowfall,month
count,8465,8465.000000,8465.000000,8465.000000,8465.000000,8465.000000,8465.000000,8465.000000,8465.000000,8465.00000,8465.000000,8465.000000
mean,2018-05-28 09:14:23.744831744,729.156999,11.507029,12.771057,58.147194,1.725883,1433.873479,3.944997,0.567868,0.14912,0.077685,6.438984
min,2017-12-01 00:00:00,2.000000,0.000000,-17.800000,0.000000,0.000000,27.000000,-30.600000,0.000000,0.00000,0.000000,1.000000
25%,2018-02-27 00:00:00,214.000000,6.000000,3.000000,42.000000,0.900000,935.000000,-5.100000,0.000000,0.00000,0.000000,3.000000
50%,2018-05-28 00:00:00,542.000000,12.000000,13.500000,57.000000,1.500000,1690.000000,4.700000,0.010000,0.00000,0.000000,6.000000
75%,2018-08-24 00:00:00,1084.000000,18.000000,22.700000,74.000000,2.300000,2000.000000,15.200000,0.930000,0.00000,0.000000,9.000000
max,2018-11-30 00:00:00,3556.000000,23.000000,39.400000,98.000000,7.400000,2000.000000,27.200000,3.520000,35.00000,8.800000,12.000000
std,NaN,642.351166,6.920899,12.104375,20.484839,1.034281,609.051229,13.242399,0.868245,1.12554,0.444063,3.452362


# **<u> Vue d'ensemble </u>**

## <u> - Question 1 : Combien de locations de vélos ont été enregistées sur l'ensemble de l'année?<u>

In [16]:
total_locations = df_active["rented_bike_count"].sum()
print(f"Volume total de locations : {total_locations:,} locations")

Volume total de locations : 6,172,314 locations


# Commentaire :
## Interprétation : Le service de vélos en libre-service a enregistré 6 172 314 locations sur l'ensemble de l'année (hors périodes de fermeture du service). Ce volume constitue un indicateur global de l'utilisation du réseau et servira de référence pour analyser les variations selon le temps, les saisons et les conditions météorologiques.


## - Comment le nombre de locations de vélos évolue-t-il au cours du temps ?

In [ ]:
# Créer le nom du mois (nécessaire pour le groupby et les graphiques)
mois_fr = {1:"Janvier",2:"Février",3:"Mars",4:"Avril",5:"Mai",6:"Juin",
           7:"Juillet",8:"Août",9:"Septembre",10:"Octobre",11:"Novembre",12:"Décembre"}
df_active["month_name"] = df_active["month"].map(mois_fr)

monthly_rentals = (df_active.groupby(["month", "month_name"])["rented_bike_count"].sum().reset_index().sort_values("month"))


# <u>Graphique recommandé pour analyse d'une évolution dans le temps c'est une courbe.</u>

In [20]:
fig = px.line(monthly_rentals, x="month_name", y="rented_bike_count", markers=True, title="Évolution mensuelle du nombre de locations de vélos",
              labels = {"month_name": "Mois",
                        "rented_bike_count": "Nombre de locations"})
fig.update_traces(
    line=dict(color="#0d6efd", width=4),
    marker=dict(
        color="#ff7f0e",
        size=9,
        line=dict(color="black",
                  width=1)),textposition="top center")

fig.update_layout(
    template="plotly_white",
    title={
        "text":"📈 Évolution mensuelle du nombre de locations de vélos",
        "x": 0.5},
    xaxis_title="Mois",
    yaxis_title="Nombre de locations",
    hovermode="x unified",
    font=dict(size=14),
    height=500)
                        
fig.show()

NameError: name 'px' is not defined

## La demande augmente progressivement entre le printemps et l'été, atteint un maximum pendant les mois les plus chauds, puis diminue à l'approche de l'hiver.
## Cette évolution suggère un effet important de la saison sur l'utilisation du service.

In [22]:
fig = px.scatter(
    monthly_rentals,
    x="month_name",
    y="rented_bike_count",
    color="rented_bike_count",
    color_continuous_scale="Turbo",
    size="rented_bike_count"
)

fig.show()

NameError: name 'px' is not defined

In [23]:
fig = px.line(
    monthly_rentals,
    x="month_name",
    y="rented_bike_count",
    text="rented_bike_count",
    markers=True,
    title="Évolution mensuelle du nombre de locations de vélos",
    labels={
        "month_name": "Mois",
        "rented_bike_count": "Nombre de locations"})

fig.update_traces(
    line=dict(color="#0d6efd", width=4),
    marker=dict(size=9),
    texttemplate="%{y:,.0f}",
    textposition="top center")

fig.update_layout(
    template="plotly_white",
    title_x=0.5,
    hovermode="x unified",
    height=500)

fig.show()

NameError: name 'px' is not defined

In [ ]:
# Construction du tableau jours ouvrés vs week-end
weekday_vs_weekend = (
    df_active.assign(**{"Type de jour": df_active["is_weekend"].map({True: "Week-end", False: "Jour ouvré"})})
    .groupby("Type de jour", as_index=False)["rented_bike_count"].mean()
)
weekday_vs_weekend


In [24]:
# 1. Préparation et calculs avancés des données
df = weekday_vs_weekend.copy()
total_loc = df["rented_bike_count"].sum()
df["part_pct"] = (df["rented_bike_count"] / total_loc) * 100

valeurs = df.set_index("Type de jour")["rented_bike_count"]
jour_max = valeurs.idxmax()
jour_min = valeurs.idxmin()
val_max = valeurs.max()
val_min = valeurs.min()

ecart_abs = val_max - val_min
ecart_pct = ((val_max - val_min) / val_min) * 100

# Calcul de l'écart relatif sur chaque barre
df["delta_vs_min"] = df["rented_bike_count"].apply(
    lambda x: f"+{ecart_pct:.1f}% vs {jour_min}"
    if x == val_max
    else "Base de comparaison"
)

# 2. Palette dynamique : couleur vive pour le leader, plus neutre pour le second
palette = {
    jour_max: "#2b5c8f",  # Bleu profond professionnel
    jour_min: "#8da0cb",}  # Bleu adouci / secondaire


# 3. Création du graphique de base
fig = px.bar(
    df,
    x="Type de jour",
    y="rented_bike_count",
    color="Type de jour",
    color_discrete_map=palette,
    custom_data=["part_pct", "delta_vs_min"],
    title="<b>Comparaison de la demande : Jours ouvrés vs Week-end</b>",
)

# 4. Enrichissement de l'interactivité au survol (Hover Custom)
fig.update_traces(
    texttemplate="<b>%{y:,.0f}</b>",
    textposition="outside",
    textfont=dict(size=14, color="#212529"),
    width=0.45,
    hovertemplate=(
        "<b>Catégorie :</b> %{x}<br>"
        "<b>Moyenne quotidienne :</b> %{y:,.0f} locations<br>"
        "<b>Part du volume total :</b> %{customdata[0]:.1f}%<br>"
        "<b>Écart :</b> %{customdata[1]}<extra></extra>"
    ),
)

# 5. Annotation style "Badge Executive" (Bannière d'accroche)
fig.add_annotation(
    x=0.5,
    y=val_max * 1.22,
    xref="paper",
    yref="y",
    text=f"🔥 <b>{jour_max}</b> surpasse <b>{jour_min}</b> de <b>+{ecart_pct:.0f}%</b> (+{ecart_abs:,.0f} locations/jour)",
    showarrow=False,
    font=dict(size=13, color="#1e293b"),
    bgcolor="#f1f5f9",
    bordercolor="#cbd5e1",
    borderwidth=1,
    borderpad=8,
)

# 6. Ligne visuelle de comparaison (Connecteur d'écart)
fig.add_shape(
    type="line",
    x0=0,
    y0=val_min,
    x1=1,
    y1=val_min,
    line=dict(color="#94a3b8", width=1.5, dash="dot"),
)

# 7. Menu interactif (Boutons pour changer de vue : Valeurs vs Parts %)
fig.update_layout(
    updatemenus=[
        dict(
            type="buttons",
            direction="left",
            x=1.0,
            y=1.18,
            xanchor="right",
            yanchor="top",
            showactive=True,
            buttons=[
                dict(
                    label="Volume Moyen",
                    method="update",
                    args=[
                        {"y": [df["rented_bike_count"]], "texttemplate": "<b>%{y:,.0f}</b>"},
                        {"yaxis": {"title": "Nombre moyen de locations par jour", "range": [0, val_max * 1.3]}},
                    ],
                ),
                dict(
                    label="Part de Marché (%)",
                    method="update",
                    args=[
                        {"y": [df["part_pct"]], "texttemplate": "<b>%{y:.1f}%</b>"},
                        {"yaxis": {"title": "Pourcentage du volume total (%)", "range": [0, 100]}},
                    ],
                ),
            ],
        )
    ],
    template="plotly_white",
    title_x=0.0,  # Titre aligné à gauche (norme moderne de Dataviz)
    title_font=dict(size=18, color="#0f172a", family="Inter, Arial"),
    showlegend=False,
    hoverlabel=dict(
        bgcolor="#ffffff",
        font_size=13,
        font_family="Arial",
        bordercolor="#cbd5e1",
    ),
    xaxis=dict(
        title="",
        showgrid=False,
        tickfont=dict(size=14, color="#334155"),
    ),
    yaxis=dict(
        title="Nombre moyen de locations par jour",
        showgrid=True,
        gridcolor="#f1f5f9",
        range=[0, val_max * 1.3],
    ),
    margin=dict(l=50, r=50, t=110, b=40),
    plot_bgcolor="white",
    paper_bgcolor="white",
)

# 8. Affichage avec barre d'outils nettoyée
fig.show(config={"displayModeBar": True, "displaylogo": False})

NameError: name 'weekday_vs_weekend' is not defined

# Commentaire :
##  On observe que : le nombre moyen de locations est plus élevé les jours ouvrés que durant le week-end ; l'écart est d'environ 1 624 locations supplémentaires par jour ; cela représente une augmentation d'environ 10 % par rapport au week-end.

## Comment la demande évolue-t-elle saison par saison ?

In [ ]:
# Construction du tableau des locations moyennes par saison
season_rentals = df_active.groupby("season", as_index=False)["rented_bike_count"].mean()
season_rentals


In [25]:
# Identifier la saison ayant la demande moyenne la plus élevée
best_season = season_rentals.loc[
    season_rentals["rented_bike_count"].idxmax()]

# Création de la courbe
fig = px.line(
    season_rentals,
    x="season",
    y="rented_bike_count",
    markers=True,
    text="rented_bike_count",
    title="<b>Évolution de la demande moyenne de locations selon les saisons</b>",
    labels={
        "season": "Saison",
        "rented_bike_count": "Nombre moyen de locations par jour"})

# Personnalisation de la courbe
fig.update_traces(
    line=dict(color="#0d6efd", width=4),
    marker=dict(
        size=10,
        color="#0d6efd",
        line=dict(color="white", width=2)
    ),
    texttemplate="%{text:,.0f}",
    textposition="top center",
    hovertemplate=(
        "<b>Saison :</b> %{x}<br>"
        "<b>Locations moyennes :</b> %{y:,.0f}<extra></extra>"))

# Annotation automatique sur la meilleure saison
fig.add_annotation(
    x=best_season["season"],
    y=best_season["rented_bike_count"],
    text="Demande maximale",
    showarrow=True,
    arrowhead=2,
    arrowsize=1,
    arrowwidth=2,
    arrowcolor="#dc3545",
    ax=0,
    ay=-45,
    bgcolor="#fff3cd",
    bordercolor="#dc3545",
    borderwidth=1,
    font=dict(size=12, color="#212529")
)

# Mise en page
fig.update_layout(
    template="plotly_white",

    title_x=0.5,
    title_font=dict(
        size=18,
        color="#212529"
    ),

    xaxis=dict(
        title="Saison",
        showgrid=False
    ),

    yaxis=dict(
        title="Nombre moyen de locations par jour",
        showgrid=True,
        gridcolor="#f1f5f9"
    ),

    hoverlabel=dict(
        bgcolor="white",
        font_size=13,
        font_family="Arial"
    ),

    font=dict(size=13),

    margin=dict(
        l=50,
        r=40,
        t=80,
        b=40
    ),

    height=500
)

fig.show()

NameError: name 'season_rentals' is not defined

## - Quelle est la demande moyenne par heure sur une journée ? 

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
# Quelle est la demande moyenne par heure sur une journée ?
# Calcul de la moyenne par heure
mean_demand_per_hour = df_active.groupby('hour')['rented_bike_count'].mean().round()

plt.figure(figsize=(10, 5))
plt.bar(mean_demand_per_hour.index, mean_demand_per_hour.values, color='limegreen', edgecolor='black')
plt.title('Demande moyenne de vélos par heure sur une journée')
plt.xlabel('Heure de la journée (0-23h)')
plt.ylabel('Nombre moyen de vélos loués')
plt.xticks(range(0, 24))
plt.grid(axis='y', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.savefig('demande_moyenne_par_heure.png')
plt.show()

print(mean_demand_per_hour)

# Commentaire :
### On constate q'un pique de location est présent à 8H mais également à 18H. Celà semble correspondre aux horaires de départ et de retour du travail.
### Cependant, un ecart assez important est constater entre les deux, celà peut peut-être s'expliquer que le moyen de transport le matin est différent du soir
### Exemple : Taxi le matin et retour en vélo le soir peut-être en raison du fort traffic a 18H. 
### Pique du matin (8h) : ~1 016 vélos en moyenne.
### Pique du soir (18h) : ~1 503 vélos en moyenne (maximum de la journée).
### Creux de la nuit (4h) : ~133 vélos en moyenne (minimum de la journée).

## - Quelle est la demande moyenne par heure sur une année ? 

In [ ]:
import matplotlib.pyplot as plt

# 1. Calcul de la demande moyenne
demande_moyenne = df_active.groupby('hour')['rented_bike_count'].mean()

# 2. Graphique en courbe
plt.figure(figsize=(10, 5))
plt.plot(demande_moyenne.index, demande_moyenne.values, color='limegreen', marker='o', linewidth=2.5)

# 3. Personnalisation
plt.title('Demande moyenne de vélos par heure sur une année')
plt.xlabel('Heure de la journée (0-23h)')
plt.ylabel('Nombre moyen de vélos loués')
plt.xticks(range(0, 24))
plt.grid(True, linestyle='--', alpha=0.7)

plt.tight_layout()
plt.show()

## - Quelle est la demande moyenne heure par heure ? 

In [ ]:
import matplotlib.pyplot as plt

demande_moyenne = df_active.groupby('hour')['rented_bike_count'].mean()

plt.figure(figsize=(8, 4))
plt.fill_between(demande_moyenne.index, demande_moyenne.values, color='skyblue', alpha=0.5)
plt.plot(demande_moyenne.index, demande_moyenne.values, color='Slateblue', linewidth=2)

plt.title('Graphique d\'aire : Demande moyenne par heure')
plt.xlabel('Heure')
plt.ylabel('Demande moyenne')
plt.grid(True, linestyle='--', alpha=0.5)
plt.show()

## - Nombre total de trajets, de vélos disponibles et de stations ?

In [ ]:
import matplotlib.pyplot as plt
# Nombre total de trajet à l'année : 
# Calcul de la métrique
total_locations = df_active['rented_bike_count'].sum()

# Création de la figure
fig, ax = plt.subplots(figsize=(6, 3))

# Ajout du texte au centre
# Formatage avec séparateur de milliers
ax.text(0.5, 0.6, f"{total_locations:,}".replace(',', ' '), 
        fontsize=50, fontweight='bold', ha='center', va='center', color='#007acc')

# Ajout du label
ax.text(0.5, 0.3, "Locations de Vélos sur un An (Total)", 
        fontsize=16, ha='center', va='center', color='#555555')

# Cacher les axes
ax.axis('off')

# Affichage
plt.tight_layout()
plt.show()

## - Comment le nombre de trajets évolue-t-il mois après mois ?

In [ ]:
# 1. Conversion de la date et groupement par mois
df_active['YearMonth'] = df_active['date'].dt.to_period('M')

monthly_trips = df_active.groupby('YearMonth')['rented_bike_count'].sum().reset_index()
monthly_trips['YearMonth_str'] = monthly_trips['YearMonth'].astype(str)

# 2. Création du graphique
fig, ax = plt.subplots(figsize=(10, 5))

# Tracé de la courbe
ax.plot(monthly_trips['YearMonth_str'], monthly_trips['rented_bike_count'], 
        color='#d9534f', marker='o', linewidth=2.5, markersize=8, zorder=3)

# Formater l'axe Y en milliers (ex: 800 k)
ax.yaxis.set_major_formatter(plt.FuncFormatter(lambda x, pos: f'{x*1e-3:,.0f} k'))

# Ajouter la valeur au-dessus de chaque point
for i, txt in enumerate(monthly_trips['rented_bike_count']):
    ax.annotate(f'{txt*1e-3:,.0f}k', 
                (monthly_trips['YearMonth_str'][i], monthly_trips['rented_bike_count'][i]),
                textcoords="offset points", 
                xytext=(0, 10), 
                ha='center', fontsize=9, fontweight='bold', color='#333333')

# Ajuster les limites de l'axe Y
ax.set_ylim(0, monthly_trips['rented_bike_count'].max() * 1.15)

# Titres et axes
ax.set_title('Évolution du nombre de trajets mois après mois', fontsize=13, fontweight='bold', pad=15)
ax.set_xlabel('Mois', fontsize=11)
ax.set_ylabel('Nombre total de trajets', fontsize=11)
ax.grid(True, linestyle='--', alpha=0.5, zorder=0)

plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

# Commentaire :
### On peut constater une augmentation du nombre de trajet par moi qui augmente la première moitié de l'année et ensuite diminue de plus en plus.
### La courbe jour par jour montre une grande variabilité (pics et chutes brusques), principalement 
### liée aux épisodes météo défavorables (pluie, neige) sur certains jours précis.

# **<u> Saisonnalité / Temporalité </u>**

### <b> Prise en compte de l’humidité avec la variable de la température dans l'utilisation des vélos </b>

In [32]:
df_active_positive = df_active[df_active['temperature'] >= 0]

df_active_positive.head()

KeyError: 'temperature'

In [33]:
df_active_negative = df_active[df_active['temperature'] < 0]

df_active_negative.head()

KeyError: 'temperature'

In [34]:
print("Température moyenne globale :", df_active['temperature'].mean())
print("Température moyenne (<0°C) :", df_active_negative['temperature'].mean())
print("Température moyenne (>=0°C) :", df_active_positive['temperature'].mean())

KeyError: 'temperature'

In [35]:
print("Humidité moyenne globale :", df_active['humidity'].mean())
print("Humidité moyenne pour les températures négatives :", df_active_negative['humidity'].mean())
print("Humidité moyenne pour les températures positives :", df_active_positive['humidity'].mean())

KeyError: 'humidity'

### Effet sur la location de vélos

In [36]:
df_humidity_final = pd.DataFrame({
    'Température moyenne': [
        df_active['temperature'].mean(),
        df_active_negative['temperature'].mean(),
        df_active_positive['temperature'].mean()],
    'Humidité moyenne': [
        df_active['humidity'].mean(),
        df_active_negative['humidity'].mean(),
        df_active_positive['humidity'].mean()],
    'Locations moyennes': [
        df_active['rented_bike_count'].mean(),
        df_active_negative['rented_bike_count'].mean(),
        df_active_positive['rented_bike_count'].mean()]}, index=['Moyenne globale', 'Température < 0°C', 'Température ≥ 0°C'])

df_humidity_final

KeyError: 'temperature'

#### **commentaire d'analyse** : on comprends qu'en moyenne, 729 vélos sont utilisés quotidiennement lorsque la température avoisine les 12°c ; la différence la plus flagrante réside entre les températures basses et élevées, avec une différence en moyenne de plus de 639 vélos utilisé quand la température est "élevée" (ici en moyenne 16.51°c).
#### L'explication externe la plus plausible est la prise en compte des températures qui influence l'humeur de l'utilisateur et donc son envie ou non d'emprunter un vélo en libre-service. Il est aussi plus propice qu'un utilisateur souhaite profiter des saisons douces et chaudes en faisant une balade à vélo que pendant la saison hivernale

## <u> Graphique 1</u>  de la prise en compte de l’humidité avec la variable de la température (diagramme barres groupées)

In [37]:
df_humidity_final[['Humidité moyenne', 'Locations moyennes']].plot(
    kind='bar',
    color= ['blue', 'green'],
    figsize=(8,5)
)

plt.title("Humidité et location des vélos selon la température")
plt.ylabel("Valeur moyenne")
plt.xticks(rotation=0)
plt.show()

NameError: name 'df_humidity_final' is not defined

#### Commentaire : graphique qui est facilement lisible, avec côte à côte l'humidité moyenne et les locations moyennes. On peut y voir clairement la différence entre les températures en dessous et au-dessus de 0.

## <u> Graphique 2</u> de la moyenne du rayonnement solaire par saison et par mois (diagramme de dispersion - interactif)

In [ ]:
# Construction de la moyenne du rayonnement solaire par mois/saison
solar_radiation_by_month = (
    df_active.groupby(["month", "month_name", "season"], as_index=False)["solar_radiation"].mean()
    .sort_values("month")
    .rename(columns={"month_name": "date"})
)
solar_radiation_by_month


In [38]:
fig = px.scatter(
    solar_radiation_by_month,
    x="date",
    y="solar_radiation",
    color="season",
    title="Moyenne du rayonnement solaire par saison et par mois")

fig.show()

NameError: name 'px' is not defined

#### commentaire : graphique lisible et compréhensible, avec une différenciation de couleurs entre chaque saison (1 point = 1 mois).

## <u> Graphique 3</u> de la moyenne du rayonnement solaire par saison et par mois (courbe temporelle - interactif)

In [39]:
fig = px.line(
    solar_radiation_by_month,
    x="date",
    y="solar_radiation",
    color="season",
    title="Moyenne du rayonnement solaire par saison et par mois")

fig.show()

NameError: name 'px' is not defined

#### commentaire : graphique beaucoup plus parlant visuellement et compréhensible. Simple incompréhension au niveau de la saison hivernale, voir s'il y a possibilité de rendre un peu plus flou le trait qui casse le graphique en plein milieu.

## <u> Graphique 5</u> de la corrélation entre le rayonnement solaire et le nombre de vélos loués durant chaque saison

In [40]:
corr = pd.DataFrame({
    "Saison": ["Automne", "Printemps", "Été", "Hiver"],
    "Corrélation": [0.245651, 0.367398, 0.088417, 0.185403]})

fig = px.bar(
    corr,
    x="Saison",
    y="Corrélation",
    color="Corrélation",
    text_auto=".2f",
    title="Corrélation entre le rayonnement solaire et le nombre de vélos loués durant chaque saison")

fig.update_layout(template="plotly_white")
fig.show()

NameError: name 'px' is not defined

## <u> Graphique 1</u> de l'influence météorologique sur la location vélos (Heatmap)

In [41]:
corr = df_active[['rented_bike_count',
           'temperature',
           'rainfall',
           'visibility',
           'wind_speed']].corr()

plt.figure(figsize=(6,5))
sns.heatmap(corr, annot=True, cmap='RdBu', vmin=-1, vmax=1)

plt.title("Corrélation entre les variables météorologiques et les locations de vélos")
plt.show()

KeyError: "None of [Index(['rented_bike_count', 'temperature', 'rainfall', 'visibility',\n       'wind_speed'],\n      dtype='object')] are in the [columns]"

## <u> Graphique 4</u> de la moyenne de locations en cas d'intempérie : neige et pluie (diagramme en barre verticale - interactif)

In [42]:
location_pluie = df_active[df_active['rainfall'] > 0]['rented_bike_count'].mean()
location_neige = df_active[df_active['snowfall'] > 0]['rented_bike_count'].mean()

df_meteo = pd.DataFrame({
    'Météo': ['Pluie', 'Neige'],
    'Locations moyennes': [location_pluie, location_neige]
})

fig = px.bar(
    df_meteo,
    x='Météo',
    y='Locations moyennes',
    color='Météo',
    title="Locations moyennes pendant un temps de pluie et de neige",
    text_auto='.0f'
)

fig.update_layout(
    yaxis_title="Nombre moyen de vélos loués",
    xaxis_title=""
)

fig.show()

KeyError: 'rainfall'

#### commentaire : graphique clair et pertinant. Ici il a peu neigé sur les 365 jours pourtant les vélos ont plus été loué par jours de neige. Cela est intéressant car il y a plus de jours ou il a plu que de jours ou il a neigé.

## - Quelles heures et quels jours sont les plus fréquentés ?

### Carte de chaleur (heatmap) : Repérer des motifs récurrents dans deux dimensions croisées (par exemple jour × heure), grâce à une intensité de couleur.

In [43]:
import plotly.express as px

jours_ordre = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]

pivot = df_active.pivot_table(
    values="rented_bike_count",
    index="hour",
    columns="day_of_week",
    aggfunc="mean"
)
pivot = pivot[jours_ordre]

fig = px.imshow(
    pivot,
    labels=dict(x="Jour de la semaine", y="Heure", color="Locations moyennes"),
    color_continuous_scale="Blues",
    title="Demande moyenne de vélos selon l'heure et le jour de la semaine",
    aspect="auto"
)
fig.show()

KeyError: 'rented_bike_count'

## - Existe-t-il une température seuil (au-delà/en-deçu) à de laquelle la demande chute ?

### Courbe de tendance :  Suivre l'évolution d'une valeur dans le temps et repérer une tendance, une saisonnalité ou une rupture 

In [44]:
import plotly.express as px

fig = px.scatter(
    df_active,
    x="temperature",
    y="rented_bike_count",
    opacity=0.1,
    trendline="lowess",
    labels={
        "temperature": "Température (°C)",
        "rented_bike_count": "Nombre de vélos loués"
    },
    title="Relation entre température et demande de vélos"
)

fig.update_traces(marker=dict(size=4))

# Cibler uniquement la courbe de tendance pour changer sa couleur
fig.update_traces(
    line=dict(color="red", width=3),
    selector=dict(mode="lines")
)

fig.show()

ValueError: Value of 'x' is not the name of a column in 'data_frame'. Expected one of ['Date', 'Rented Bike Count', 'Hour', 'Temperature(°C)', 'Humidity(%)', 'Wind speed (m/s)', 'Visibility (10m)', 'Dew point temperature(°C)', 'Solar Radiation (MJ/m2)', 'Rainfall(mm)', 'Snowfall (cm)', 'Seasons', 'Holiday', 'Functioning Day', 'YearMonth'] but received: temperature

### Commentaire : 
#### Le nuage de points montre une tendance claire : plus la température augmente, plus la demande de vélos augmente. 
#### Mais la courbe de tendance (en rouge) révèle que cette relation n’est pas parfaitement linéaire, et qu’elle se décompose en trois phases distinctes :
#### 1.	En dessous de 0°C : la demande reste très faible et proche de zéro, quasi plate. Le froid semble décourager fortement l’usage du vélo.
#### 2.	Entre 0°C et 25°C environ : la courbe grimpe de façon soutenue, c’est la zone où chaque degré supplémentaire fait le plus gagner en demande.
#### 3.	Au-delà de 25°C : la courbe s’aplatit nettement. La demande continue de progresser très légèrement, mais beaucoup plus lentement comme si un plafond de confort était atteint.

## - Comment l'usage varie-t-il selon les saisons ou la météo ?


### Boîte à moustaches (box plot) : Visualiser la distribution d'une variable (médiane, dispersion, valeurs extrêmes) et comparer plusieurs groupes.

In [45]:
# Ordonner les saisons logiquement
saisons_ordre = ["Spring", "Summer", "Autumn", "Winter"]

plt.figure(figsize=(10, 6))
sns.boxplot(
    data=df_active,
    x="season",
    y="rented_bike_count",
    order=saisons_ordre,
    palette="YlOrRd"
)
plt.title("La demande varie fortement selon les saisons")
plt.xlabel("Saison")
plt.ylabel("Nombre de vélos loués")
plt.show()

ValueError: Could not interpret value `season` for `x`. An entry with this name does not appear in `data`.

<Figure size 1000x600 with 0 Axes>

## Commentaire :
#### Ce qu’on observe par saison
#### •	Été et Automne ont la médiane la plus haute (respectivement 950 et 800 vélos/heure) et les boîtes les plus larges 
####    la demande y est non seulement forte, mais aussi la plus variable d’une heure à l’autre.
#### •	Printemps se situe en position intermédiaire, avec une médiane autour de 550.
#### •	Hiver se distingue nettement : médiane basse (200) et boîte très resserrée 
####    la demande hivernale est à la fois faible et peu changeante.
#### la comparaison été/automne vs hiver, et l’explication rapide des ronds noirs (heures exceptionnelles de forte affluence).

# - La pluie ou la neige ont-elles un effet plus fort que la température sur la baisse de la demande ?

## Barres comparatives : 

In [46]:
import pandas as pd
import plotly.express as px

def categorie_meteo(row):
    if row["snowfall"] > 0:
        return "Neige"
    elif row["rainfall"] > 0:
        return "Pluie"
    else:
        return "Sec"

df_active["meteo"] = df_active.apply(categorie_meteo, axis=1)

df_active["temp_bin"] = pd.cut(
    df_active["temperature"],
    bins=[-20, 0, 15, 40],
    labels=["Froid (<0°C)", "Doux (0-15°C)", "Chaud (>15°C)"]
)
resume = df_active.groupby(["temp_bin", "meteo"], observed=True)["rented_bike_count"].mean().reset_index()

color_discrete_map={
    "Sec": "#F4A460",      # sable/beige doré - évoque le soleil, temps sec
    "Pluie": "#4682B4",    # bleu acier - évoque l'eau, la pluie
    "Neige": "#B0C4DE",    # bleu très pâle/blanc bleuté - évoque la neige/glace
}

fig = px.bar(
    resume,
    x="temp_bin",
    y="rented_bike_count",
    color="meteo",
    barmode="group",
    color_discrete_map=color_discrete_map,
    labels={
        "temp_bin": "Tranche de température",
        "rented_bike_count": "Nombre moyen de vélos loués",
        "meteo": "Météo"
    },
    title="La pluie réduit fortement la demande à toute température ; la neige, plus rare, semble avoir un effet moindre en conditions douces"
)
fig.show()

KeyError: 'snowfall'

## Commentaire :
### Chaud : la barre “Sec” est de loin la plus haute de tout le graphique beau temps chaud = pic de demande.
### Doux : “Sec” domine largement, “Pluie” chute très fort (la plus basse barre affichée), et “Neige” est étonnamment plus haute que Pluie dans cette tranche.
### Froid : les barres “Sec” et “Neige” sont proches l’une de l’autre, sans grand écart.

In [ ]:
df_active.groupby(["temp_bin", "meteo"]).size()